build-tools/src/main/java/org/elasticsearch/gradle/plugin/GenerateTestBuildInfoTask.java

```diff
@@ -0,0 +1,294 @@
+
+    @Input
+    public abstract Property<String> getComponentName();
+
+    @InputFiles
+    public abstract Property<FileCollection> getCodeLocations();
+
+    @Input
+    public abstract Property<String> getOutputFileName();
+
+    @OutputDirectory
+    public abstract DirectoryProperty getOutputDirectory();
+
+    @TaskAction
+    public void generatePropertiesFile() throws IOException {
+        Map<String, String> classesToModules = buildClassesToModules();
+
+        Path outputDirectory = getOutputDirectory().get().getAsFile().toPath();
+        Files.createDirectories(outputDirectory);
+        Path outputFile = outputDirectory.resolve(getOutputFileName().get());
+
+        try (var writer = Files.newBufferedWriter(outputFile, StandardCharsets.UTF_8)) {
+            writer.write("{\n");
+
+            writer.write("    \"name\": \"");
+            writer.write(getComponentName().get());
+            writer.write("\",\n");
+
+            writer.write("    \"locations\": [\n");
+            if (classesToModules.isEmpty() == false) {
+                StringBuilder sb = new StringBuilder();
+                for (Map.Entry<String, String> entry : classesToModules.entrySet()) {
+                    sb.append("        {\n");
+                    sb.append("            \"class\": \"");
+                    sb.append(entry.getKey());
+                    sb.append("\",\n            \"module\": \"");
+                    sb.append(entry.getValue());
+                    sb.append("\"\n        },\n");
+                }
+                writer.write(sb.substring(0, sb.length() - 2));
+            }
+            writer.write("\n    ]\n}\n");
+        }
+    }
+
+    // build the association for a class to a module;
+    // there are different methods for finding these depending on if the
+    // classpath entry is a jar or a directory
+    private Map<String, String> buildClassesToModules() throws IOException {
+        Map<String, String> classesToModules = new HashMap<>();
+        for (File file : getCodeLocations().get().getFiles()) {
+            if (file.exists()) {
+                if (file.getName().endsWith(".jar")) {
+                    extractFromJar(file, classesToModules);
+                } else if (file.isDirectory()) {
+                    extractFromDirectory(file, classesToModules);
+                } else {
+                    throw new IllegalArgumentException("unrecognized classpath entry: " + file);
+                }
+            }
+        }
+        return classesToModules;
+    }
+
+    // find the first class and module when the class path entry is a jar
+    private void extractFromJar(File file, Map<String, String> classesToModules) throws IOException {
+        try (JarFile jarFile = new JarFile(file)) {
+            String className = extractClassNameFromJar(jarFile);
+            String moduleName = extractModuleNameFromJar(file, jarFile);
+
+            if (className != null) {
```

Comment:
Do we have instances of jars with no classes?

Reply:
I ran into this, but I don't remember specifically which jar it was now.